report/3.Theory/TypeRules.tex

```diff
@@ -1,13 +1,373 @@
+x = 10 + (2 equals 3)
+\end{lstlisting}
+
+To avoid this, the HCL compiler includes a type-checker which will be described in a later chapter.
+
+The HCL type system consists of three components:
+\begin{itemize}
+	\item The set of all types in HCL.
+	\item Type rules for each syntactic category.
+	\item A definition for each type rule.
+\end{itemize}
+
+Essentially, the type rules specify when certain expressions have to evaluate to specific types.
+The rules have been split up in two sections:
+\begin{itemize}
+	\item Rules for expressions
+	\item Rules for statements
+\end{itemize}
+The former specifies which type an expression evaluates to, depending on the input.
+The latter specifies what types the components of a statement must be to be type valid.
+
+\textbf{Notation}\\
+The notation for type rules is very similar to the notation of semantics with a few differences.
+Below is an example:\\
+\begin{center}
+	\begin{math}
+	\cfrac
+	{E \vdash e_1 : num \quad e_2 : num}
+	{E \vdash e_1\ + e_2 : num}
+	\end{math}
+\end{center}
+This is the type rule for addition of two numbers.
+It means that given the current environment, the resulting value will be of type $num$ if the two input expressions are both of type $num$.
+
+As the type rules do not change the state of a program, and therefore also do not change the variable- and function environments, the environment can be simplified to one symbol, $E$.
+The colon is used to show that an expression is of a certain type.
 
 \subsection{Type Rules for Expressions}
-%quick introduction.
+As mentioned earlier, an expression evaluates to a type, if certain premises are upheld (where such apply).
+For the sake of brevity, the type rules for expressions have been simplified significantly compared to the semantic rules.
+Unlike the semantics, where certain function calls must have its own rule in order for the semantics of HCL to be properly defined, the type rules for all function calls can be explained with a single rule.
+The following are the type rules for all expressions.
+
+\textbf{Variable type}\\
+The type of a variable is found by looking up the variable in the environment:\\
+\begin{center}
+	\begin{math}
+	E(x) : T
+	\end{math}
+	\\[1\baselineskip]
+	\texttt{where} $T$ is the type that the variable was declared with
+\end{center}
+
+
+\textbf{Num literal type}\\
+A $num$ literal is always of type $num$.
+
+\begin{center}
+	\begin{math}
+		E \vdash n : num
+	\end{math}
+\end{center}
+
+\textbf{Bool literal type}\\
+A $bool$ literal is always of type $bool$.
+
+\begin{center}
+	\begin{math}
+	E \vdash b : bool
+	\end{math}
+\end{center}
+
+\textbf{Txt literal type}\\
+A $txt$ literal is always of type $txt$.
+
+\begin{center}
+	\begin{math}
+	E \vdash t : txt
+	\end{math}
+\end{center}
+
+\textbf{List literal type}\\
+Every entry in any given list must be of the same type, as long as that type isn't $none$.
```

Comment:
"but cannot be of type none".. the andet implyer at du godt må have forskellige typer hvis den ene type er none